### PR TITLE
TF-TRT C++ conversion

### PR DESCRIPTION
**Pull request by tfeher**
TF-TRT C++ interface to convert models. 

Differences compared to the Python API:
- It is assumed that the input graph is frozen. This assumption will be removed in a follow up PR.
- `convert_to_static_engine` conversion param allows to convert dynamic engines to static engines.
- Currently we do not provide a way to save the engine files as assets. 

The steps for model conversion by `ConvertAndBuild`
1. Inline functions
2. Freeze the graph - omitted since we assume frozen input graph
3. Run Grappler with TRTOptimizer pass
4. Infer the graph to provide shape information (only in dynamic shape mode).
5. Infer the graph to build the engines
6. Convert the graph_def to have static engines

(On the Python side, steps 4-5 are done by a separate `build` function.)

Related PRs:
- #52047 
- Example code for the tensorflow/tensorrt repository https://github.com/tensorflow/tensorrt/pull/271

### CODE REVIEW

**Comment by bixia1** [comment]
We can avoid the change to the utilities for freezing the graph by updating out SavedModelBundle with the converted GraphDef. Something like this:
```
MetaGraphDef* meta_graph_def = &saved_model_bundle->meta_graph_def; 
*meta_graph_def->mutable_graph_def() = graph_def;
```

**Comment by tfeher** [comment]
@bixia1 ready for review.

File: tensorflow/core/BUILD [hunk 1]
```diff
@@ -748,6 +748,7 @@ tf_cuda_library(
         ":gpu_runtime",
         ":lib",
         ":ops",
+        "//tensorflow/compiler/tf2tensorrt:trt_api"
```
Comment (tfeher) [inline tensorflow/core/BUILD:751]: @bixia1 could you advise a better place to link this in the build tree?

File: tensorflow/compiler/tf2tensorrt/trt_convert.h [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  // otherwise native TF is used. We recommend to set it false and build the
+  // engine in advance, to avoid runtime overhead.
+  bool allow_build_at_runtime = true;
+
+  // Record the TRT engine as an attribute for the TRTEngineOp. This is only
+  // valid when max_cached_engines == 1. Note: the frozen graph together with
+  // the serialized engines have to be below 2GiB (protobuf size limit). If
+  // convert_to_static_engine = false, then the converted graph_def only
+  // contains placeholder TRTEngineOp nodes, and the TRT engines will be created
+  // when we run the
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/trt_convert.h:94]: Shall we still build the cuda engine, but not convert to attribute, because the API is called ConvertAndBuild?

File: tensorflow/compiler/tf2tensorrt/trt_convert.h [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+ * Arguments:
+ * - frozen_graph_def input graph, it is assumed to be frozen.
+ * - signature
+ * - conv_params parameters for the TF-TRT conversion
+ * - inputs Graph inputs that we will use for building the TRT engines.
+ *
+ * Returns the converted graph_def.
+ */
+StatusOr<GraphDef> ConvertAndBuild(
+    const GraphDef& frozen_graph_def, const SignatureDef& signature,
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/trt_convert.h:112]: We may want to replace SignatureDef with 
std::vector<string>& input_names, const std::vector<string>& output_names

File: tensorflow/compiler/tf2tensorrt/trt_convert_test.cc [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+  std::vector<std::vector<int64>> input_shapes;
+};
+
+class TrtConverterTest : public ::testing::TestWithParam<TestParam> {
+ protected:
+  TrtConverterTest() { param_ = GetParam(); }
+
+  // Returns the following graph: output = input * [42, 137] + input
+  MetaGraphDef GetModel() {
+    // std::vector<int64> dims{-1, 2};
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/trt_convert_test.cc:42]: Not needed.

File: tensorflow/compiler/tf2tensorrt/trt_convert_test.cc [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+    PartialTensorShape shape({-1, 2});
+    // TF_CHECK_OK(TensorShapeUtils::MakeShape(dims, &shape));
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/trt_convert_test.cc:44]: Not needed.

File: tensorflow/compiler/tf2tensorrt/trt_convert_test.cc [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+        .mutable_tensor_shape()
+        ->CopyFrom(shape_proto);
+    (*out.mutable_signature_def())["serving_default"] = signature_def;
+
+    VLOG(2) << signature_def.DebugString();
+    return out;
+  }
+
+  void CheckTrtNode(const GraphDef& converted_graph_def) {
+    // Confirm that we have a TRT node with the correct attributes
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/trt_convert_test.cc:76]: Shall we move put this as a comment for the function?
Add a period to the end of the line, to be consistent with the rest.

File: tensorflow/compiler/tf2tensorrt/trt_convert_test.cc [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+      std::vector<Tensor> trt_outputs;
+      TF_EXPECT_OK(trt_session->Run({{"input", input.at(0)}}, {"output"}, {},
+                                    &trt_outputs));
+      std::cout << trt_outputs.at(0).DebugString() << std::endl;
+      ASSERT_EQ(outputs.size(), 1);
+      ASSERT_EQ(trt_outputs.size(), 1);
+      tensorflow::test::ExpectEqual(outputs[0], trt_outputs[0]);
+    }
+
+    CheckTrtNode(converted_graph_def);
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/trt_convert_test.cc:140]: Let's move this to 112, after ConvertAndBuild.

File: tensorflow/compiler/tf2tensorrt/trt_convert.cc [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+  }
+  if (!p.allow_build_at_runtime && n_inputs == 0) {
+    LOG(WARNING)
+        << "TRT will not be used since allow_build_at_runtime is disabled and "
+           "no inputs are provided to build during conversion.";
+  }
+  return Status::OK();
+}
+
+StatusOr<GraphDef> ConvertAndBuild(
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/trt_convert.cc:347]: Only this routine needs to stay in this namespace, all the rest should be in "anonymous namespace", right?

**Review by bixia1 — COMMENTED** [review]
I am still working on reviewing trt_convert.cc .

File: tensorflow/compiler/tf2tensorrt/trt_convert.cc [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+#include "tensorflow/core/grappler/optimizers/meta_optimizer.h"
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/platform/errors.h"
+#include "tensorflow/core/protobuf/meta_graph.pb.h"
+
+namespace tensorflow {
+
+namespace tensorrt {
+
+// Removes the tensor suffix after the node name (e.g. Input:0 -> Input).
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/trt_convert.cc:39]: Shall we remove parameter "prefix" as it is not used by the callsites?
If not, we need to update the document for the routine to add the description related to "prefix".

File: tensorflow/compiler/tf2tensorrt/trt_convert.cc [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+    return errors::Internal("Engine not found for", node.name());
+  }
+
+  // Serialize the engine.
+  engine_data->reset(engine->cuda_engine->serialize());
+
+  return Status::OK();
+}
+
+// Saves the TRT engines a static attribute of the TRTEngineOp nodes.
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/trt_convert.cc:292]: s/a static attribute/as attributes/

File: tensorflow/compiler/tf2tensorrt/trt_convert.cc [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+              << ", device: " << node.device();
+      TrtUniquePtrType<nvinfer1::IHostMemory> engine_data;
+      TF_RETURN_IF_ERROR(ReadSerializedEngine(node, session, &engine_data));
+      auto* attr = node.mutable_attr();
+      AttrValue static_engine;
+      static_engine.set_b(true);
+      AttrValue engine_string;
+      engine_string.set_s(engine_data->data(), engine_data->size());
+      (*attr)["static_engine"] = static_engine;
+      (*attr)["serialized_segment"] = engine_string;
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/trt_convert.cc:310]: Ideally, we should name these constants. But we can do it in another PR.

File: tensorflow/compiler/tf2tensorrt/trt_convert.cc [hunk 12]
```diff
@@ -0,0 +1,389 @@
...
+  TF_RETURN_IF_ERROR(GetEngineCacheResource(node, session, &resource));
+  core::ScopedUnref unref_cache_res(resource);
+  if (resource->cache_.size() > 1) {
+    return errors::Internal(
+        "Multiple engines found, but we can only serialize one");
+  }
+  const std::unique_ptr<EngineContext>& engine =
+      resource->cache_.begin()->second;
+  if (!engine || !engine->cuda_engine) {
+    // TODO(tfeher): Engine could have failed to build. Handle that correctly.
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/trt_convert.cc:282]: What will the "TODO" be doing?
If the TRTEngineOp is not executed, we can't build the engine. 
I think we may just issue a warning and skip the TRTEngineOp?

**Review by bixia1 — COMMENTED** [review]
Still working on trt_convert.cc.

File: tensorflow/compiler/tf2tensorrt/trt_convert.cc [hunk 13]
```diff
@@ -0,0 +1,389 @@
...
+}
+
+// Runs the model to create the engines. In dynamic shape mode, before creating
+// the engines, we provide shapes to define optimization profiles.
+Status Build(GraphDef& segmented_graph_def,
+             const std::vector<std::vector<tensorflow::Tensor>>& inputs,
+             const SignatureDef& signature, Session* session,
+             const TfTrtConversionParams params) {
+  VLOG(2) << "Building the model";
+  bool profile_generation_mode = params.use_dynamic_shape && inputs.size() > 1;
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/trt_convert.cc:212]: Let's rename profile_generation_mode to need_shape_profile.
why inputs.size() > 1? Do you mean >=1?
Also, we don't call Build if there is no input. So what we need is just this:
bool need_shape_profile = params.use_dynamic_shape

File: tensorflow/compiler/tf2tensorrt/trt_convert.cc (hunk 11 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/trt_convert.cc:310]: Let us do it in the follow up PR.

**Review by tfeher — COMMENTED** [review]
Thanks @bixia1 for the review I have addressed most of the issues. Pleas have a look.

**Comment by bixia1** [comment]
Can you fix the PR description, such as to remove the information that is obsolete? The PR description will become part of the commit message.

**Review by tfeher — COMMENTED** [review]
Thanks @bixia1 for the review, I have addressed the issues.

**Review by bixia1 — APPROVED** [review]
